### PR TITLE
Rollbacks for deadlocks

### LINKED ISSUE #30752: Deadlocks in nested transactions attempt rollback
### Steps to reproduce

If a MySQL deadlock occurs (thus raising an `ActiveRecord::Deadlocked` exception), inside a
nested transaction (i.e. a `SavepointTransaction`), Rails' rescue handler attempts to
roll back to the most recent savepoint (calling [rollback_transaction](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L233), [rollback](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L221), and then [exec_rollback_to_savepoint](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/savepoints.rb#L15)).

However, after a deadlock, the MySQL [server has already rolled back](https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlock-detection.html) the entire transaction, including any savepoints:

> InnoDB automatically detects transaction deadlocks and rolls back a transaction or transactions to break the deadlock.

This means when the "ROLLBACK TO SAVEPOINT" query is sent, the savepoint is never valid, and always raises an unnecessary and incorrect [Mysql2::Error](https://github.com/brianmario/mysql2/blob/58d0a0d4cf59/lib/mysql2/error.rb#L4) with the message "SAVEPOINT active_record_n does not exist: ROLLBACK TO SAVEPOINT active_record_n".

A test file appears below.

### Expected behavior
When a MySQL deadlock happens inside a nested `within_new_transaction(:requires_new => true)` block, `within_new_transaction` should cleanly roll back and raise an `ActiveRecord::Deadlocked` exception.

### Actual behavior
Nested `within_new_transaction(:requires_new => true)` raises a `Mysql2::Error` exception.

### System configuration
**Rails version**: seen in production 5.0.2 (with slightly different symptoms), local testing on master (c9527652c753).

**Ruby version**: production 2.3.3/2.3.5, local testing 2.4.2

### Proposed fix

From my understanding of the Rails `Transaction` and `TransactionState` classes, I think the best fix would be to introduce a new [`TransactionState`](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L5) (maybe `:invalidated`?). This state would be applied to each item in the `TransactionManager`'s stack as soon as `ActiveRecord::Deadlocked` is [rescued](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L231).

When a transaction's state is `:invalidated`, each `Transaction` subclass would change the behavior of `#rollback` in the minimal possible way ([here](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L147) and [here](https://github.com/rails/rails/blob/c9527652c753/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L170)), by omitting the `connection.rollback_*` call but continuing with all other behavior including rolling back records (right?).

Additionally, they would change the behavior of `#commit` to raise a (different) error (a commit during an exception handler would be ill-advised and probably already raise an exception, but we can at least make the language of that exception more precise).

This same error probably also happens with Postgres. My guess is that `SerializationFailure` should be handled this way exactly as `Deadlocked` is. I haven't tried it.

### Test case

This is a simple modification of the existing [`Mysql2TransactionTest`](https://github.com/rails/rails/blob/c9527652c753/activerecord/test/cases/adapters/mysql2/transaction_test.rb).

```
ubuntu@rails-dev-box:/vagrant/rails$ cat activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb
# frozen_string_literal: true

require "cases/helper"
require "support/connection_helper"

module ActiveRecord
  class Mysql2NestedDeadlockTest < ActiveRecord::Mysql2TestCase
    self.use_transactional_tests = false

    class Sample < ActiveRecord::Base
      self.table_name = "samples"
    end

    setup do
      @abort, Thread.abort_on_exception = Thread.abort_on_exception, false

      @connection = ActiveRecord::Base.connection
      @connection.clear_cache!

      @connection.transaction do
        @connection.drop_table "samples", if_exists: true
        @connection.create_table("samples") do |t|
          t.integer "value"
        end
      end

      Sample.reset_column_information
    end

    teardown do
      @connection.drop_table "samples", if_exists: true

      Thread.abort_on_exception = @abort
    end

    test "deadlock correctly raises Deadlocked inside nested SavepointTransaction" do
      assert_raises(ActiveRecord::Deadlocked) do
        barrier = Concurrent::CyclicBarrier.new(2)

        s1 = Sample.create value: 1
        s2 = Sample.create value: 2

        begin
          thread = Thread.new do
            # Start a RealTransaction
            Sample.transaction(:requires_new => false) do
              # Start a SavepointTransaction inside it
              Sample.transaction(:requires_new => true) do
                s1.lock!
                barrier.wait
                s2.update_attributes value: 1
              end
            end
          end

          begin
            # Start a RealTransaction
            Sample.transaction(:requires_new => false) do
              # Start a SavepointTransaction inside it
              Sample.transaction(:requires_new => true) do
                s2.lock!
                barrier.wait
                s1.update_attributes value: 2
              end
            end
          ensure
            thread.join
          end

        rescue ActiveRecord::StatementInvalid => e
          if /SAVEPOINT active_record_. does not exist: ROLLBACK TO SAVEPOINT/ =~ e.to_s
            assert nil, "ROLLBACK TO SAVEPOINT query issued for savepoint that no longer exists due to deadlock: #{e}"
          else
            raise e
          end
        end
      end
    end

  end
end
```

### PR DESCRIPTION
**Pull request by jamiemccarthy**
### Summary of the issue

ActiveRecord offers nested transactions, when `requires_new: true` is passed to `ActiveRecord::ConnectionAdapters::DatabaseStatements#transaction`.

In MySQL, this is not simply a matter of nesting the `BEGIN` and `COMMIT`. Instead, ActiveRecord creates a nested `SavepointTransaction` with its method `create_savepoint`, and rolls it back if necessary with `exec_rollback_to_savepoint`.

As described in #30752 , currently a deadlock inside a nested MySQL transaction results in ActiveRecord invoking `exec_rollback_to_savepoint` in an attempt to undo the inner transaction(s), before proceeding to the outermost transaction where a simple `ROLLBACK` is issued.

None of these SQL statements are necessary. By the time ActiveRecord sees the `ActiveRecord::Deadlocked` error, as far as the MySQL server is concerned, the transaction is no longer valid and none of its savepoints exist. It so happens that the server tolerates a `ROLLBACK` query (it's a no-op since the client is no longer in a transaction).

But the `ROLLBACK TO SAVEPOINT` is an error. The server emits `ERROR 1305 (42000): SAVEPOINT foo does not exist`, which ActiveRecord sees as a `Mysql2::Error` exception.

Because ActiveRecord handles exceptions within nested transactions by rolling back each savepoint in turn, a deadlock inside a nested transaction raises this unnecessary error 100% of the time.

This is particularly irksome because deadlocks are not the application's "fault"; they are a quirk of ACID design and [can happen](https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlocks-handling.html) to almost any nontrivial transaction under the wrong circumstances.

### Summary of this change

This pull request adds an additional state to `TransactionState` which indicates that the transaction has been `:invalidated`. As the `ActiveRecord::Deadlocked` error percolates to the outermost `transaction`, each state on the stack is invalidated in turn.

With this PR, an invalidated transaction continues to invoke [after_rollback callbacks](https://github.com/rails/rails/blob/9befc197f926/activerecord/lib/active_record/transactions.rb#L176-L189) as each nested transaction unwinds. It simply omits the `exec_rollback_to_savepoint` call, so it doesn't ask the server to roll back a savepoint that no longer exists. Thus the unnecessary error is avoided.

A test is provided that forces a deadlock inside a nested transaction. It confirms that the exception raised out of that code is no longer `ActiveRecord::StatementInvalid` but rather `ActiveRecord::Deadlocked`. The test passes in this branch and fails in Rails `master`. (I pushed the test commit first and got Travis to put [that failure](https://travis-ci.org/rails/rails/builds/289595574) on the record.)

### Other Information

After a transaction is invalidated, this PR also skips the final `exec_rollback_db_transaction`, because there is no point to issuing a no-op `ROLLBACK`. (It wouldn't really hurt anything, just waste a millisecond for the query.)

Additionally, with this PR, just in case some exception handler or callback tries to `commit` after a transaction has deadlocked, a `ActiveRecord::StatementInvalid` is raised. This would have been a serious but probably silent logic error otherwise.

### Commentary on alternatives to the new state

The approach I decided to take was adding a `TransactionState`, because my general feeling is that if a state is being stored in an object, the best design is to model the real-world state accurately.

Another alternative would have been to trust that the `ActiveRecord::Deadlocked` exception percolates to the outermost transaction (which is a reasonable assumption), without being rescued by any application which then naively executes code that tries to issue a query (which may or may not be a reasonable assumption). If those assumptions are accurate, one could instead avoid the `exec_rollback_db_transaction` at each nested transaction level "individually."

In my view, the main argument for that alternative design would be that my new `TransactionState` doesn't actually percolate up to the top level immediately _anyway_ (each level only tracks its children not its parent), so my implementation doesn't really fix all the states immediately in any case, so why bother fixing any of them.

To this, I would point out that with the new state it's easier to raise an exception with [a clearer error message](https://github.com/rails/rails/pull/30921/commits/c043412d149eff7a3eed7e85ffa04c628d47fe94#diff-52c8dd8e01c039c37b33b3fcbdfe406aR70).

### Commentary on the Postgresql test

I don't personally use Postgresql nor SQLite and can't say I'm terribly familiar with them. I attempted to write a test for Postgres that confirmed correct behavior both of `SerializationFailure` and `Deadlocked` in a nested transaction just like the existing `PostgresqlTransactionTest` does in a non-nested transaction.

As Travis [verifies](https://travis-ci.org/rails/rails/builds/289595574), in 1e8c7654483d38c6dbfa04e79978fbae247ad50d, both my new Postgres tests pass against the existing `master` branch, which makes that not a very useful test.

I believe that after the Postgres server invalidates the transaction, either the client must be resetting the connection's status or the server effectively does that automatically. Either way, the test is left in a state where, while commands can be issued, the client thinks it's in a transaction but the server disagrees. This could probably be tested better.

Still, I think having a test that verifies correct Rails behavior is at least a good jumping-off point to one that also errors on incorrect behavior, so here it is anyway.

I did not attempt to do the same with the sqlite3 adapter but would be open to trying.

### Commentary on invoking after_rollback callbacks

I'm not an expert on the callbacks offered for transactions, either after commit or after rollback. This change deliberately continues to invoke the after-rollback callbacks, even though the connection is in a state where any queries will raise an exception.

I believe this is the correct thing to do because any such callback should already be assuming that the connection is invalid.

I'm reasonably confident I understand the logic here, but more eyes on that design point specifically would of course be appreciated.

### CODE REVIEW

**Comment by jamiemccarthy** [comment]
(waiting for Travis to analyze the tests against existing codebase before pushing remaining commits... not sure if that's necessary but what the heck)

**Comment by jamiemccarthy** [comment]
For the record, I've pushed 21a611a1783e9c0aee97581ee5b3461e2f925dfc, but it seems wrong to me. If we were to draw the state diagram for that class, I don't think that would be a valid transition. So I would propose [deleting this test](https://github.com/voxmedia/rails/blob/f682c2ed270a4302f6813a3994920db8b6869c5c/activerecord/test/cases/transactions_test.rb#L857-L864) then reverting 21a611a1783e9c0aee97581ee5b3461e2f925dfc.

But I realize that there might be existing code that that would break, so I'm open to listening to the case for keeping the test.

**Comment by jamiemccarthy** [comment]
cc @kamipo @rafaelfranca  ... any thoughts on deadlock rollbacks?

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 1]
```diff
@@ -230,9 +258,11 @@ def within_new_transaction(options = {})
             yield
           rescue Exception => error
             if transaction
+              transaction.state.invalidate! if error.is_a? ActiveRecord::TransactionRollbackError
```
Comment (jeremy) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:261]: Do deadlocks invalidate savepoints on all databases? Will this need to be available for adapters to control behavior?

**Review by jeremy — COMMENTED** [review]
Really nice. Apologies for letting this linger. 🙌🏼

**Comment by rafaelfranca** [comment]
Merged in 7eeab35

**Pull request closed without merging**